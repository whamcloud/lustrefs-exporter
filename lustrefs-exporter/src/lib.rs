// Copyright (c) 2024 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

pub mod brw_stats;
pub mod host;
pub mod jobstats;
pub mod llite;
pub mod lnet;
pub mod openmetrics;
pub mod quota;
pub mod service;
pub mod stats;

use axum::{
    http::{self, StatusCode},
    response::{IntoResponse, Response},
};
use lustre_collector::{LustreCollectorError, TargetVariant};

#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error(transparent)]
    Http(#[from] http::Error),
    #[error(transparent)]
    TaskJoin(#[from] tokio::task::JoinError),
    #[error(transparent)]
    Io(#[from] std::io::Error),
    #[error(transparent)]
    LustreCollector(#[from] LustreCollectorError),
    #[error(transparent)]
    Utf8(#[from] std::str::Utf8Error),
    #[error("Could not find match for {0} in {1}")]
    NoCap(&'static str, String),
    #[error(transparent)]
    Otel(#[from] opentelemetry_sdk::metrics::MetricError),
}

impl IntoResponse for Error {
    fn into_response(self) -> Response {
        tracing::warn!("{self}");

        StatusCode::INTERNAL_SERVER_ERROR.into_response()
    }
}

trait LabelProm {
    fn to_prom_label(&self) -> &'static str;
}

impl LabelProm for TargetVariant {
    fn to_prom_label(&self) -> &'static str {
        match self {
            TargetVariant::Ost => "ost",
            TargetVariant::Mgt => "mgt",
            TargetVariant::Mdt => "mdt",
        }
    }
}
