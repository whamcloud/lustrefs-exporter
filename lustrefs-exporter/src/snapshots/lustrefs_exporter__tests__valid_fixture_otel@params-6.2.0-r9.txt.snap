---
source: lustrefs-exporter/src/lib.rs
expression: x
input_file: lustre-collector/src/fixtures/valid/params-6.2.0-r9.txt
---
# HELP lustre_health_healthy Indicates whether the Lustre server is healthy or not. 1 is healthy, 0 is unhealthy.
# TYPE lustre_health_healthy gauge
lustre_health_healthy{otel_scope_name="lustre"} 0
# HELP lustre_lnet_mem_used Gives information about Lustre LNet memory usage.
# TYPE lustre_lnet_mem_used gauge
lustre_lnet_mem_used{otel_scope_name="lustre"} 67130162
# HELP lustre_mem_used Gives information about Lustre memory usage.
# TYPE lustre_mem_used gauge
lustre_mem_used{otel_scope_name="lustre"} 243927005
# HELP lustre_mem_used_max Gives information about Lustre maximum memory usage.
# TYPE lustre_mem_used_max counter
lustre_mem_used_max{otel_scope_name="lustre"} 245019901
# HELP lustre_ldlm_canceld_stats Gives information about LDLM Canceld service.
# TYPE lustre_ldlm_canceld_stats counter
lustre_ldlm_canceld_stats{operation="ldlm_cancel",otel_scope_name="lustre"} 978
lustre_ldlm_canceld_stats{operation="req_active",otel_scope_name="lustre"} 978
lustre_ldlm_canceld_stats{operation="req_qdepth",otel_scope_name="lustre"} 978
lustre_ldlm_canceld_stats{operation="req_timeout",otel_scope_name="lustre"} 978
lustre_ldlm_canceld_stats{operation="req_waittime",otel_scope_name="lustre"} 978
lustre_ldlm_canceld_stats{operation="reqbuf_avail",otel_scope_name="lustre"} 1974
# HELP lustre_ldlm_cbd_stats Gives information about LDLM Callback service.
# TYPE lustre_ldlm_cbd_stats counter
lustre_ldlm_cbd_stats{operation="ldlm_bl_callback",otel_scope_name="lustre"} 134
lustre_ldlm_cbd_stats{operation="req_active",otel_scope_name="lustre"} 134
lustre_ldlm_cbd_stats{operation="req_qdepth",otel_scope_name="lustre"} 134
lustre_ldlm_cbd_stats{operation="req_timeout",otel_scope_name="lustre"} 134
lustre_ldlm_cbd_stats{operation="req_waittime",otel_scope_name="lustre"} 134
lustre_ldlm_cbd_stats{operation="reqbuf_avail",otel_scope_name="lustre"} 277
# HELP lustre_inodes_free The number of inodes (objects) available.
# TYPE lustre_inodes_free gauge
lustre_inodes_free{component="mdt",target="fs-MDT0000",otel_scope_name="lustre"} 1885250
lustre_inodes_free{component="mgt",target="MGS",otel_scope_name="lustre"} 32555
lustre_inodes_free{component="ost",target="fs-OST0000",otel_scope_name="lustre"} 40134
lustre_inodes_free{component="ost",target="fs-OST0001",otel_scope_name="lustre"} 40134
# HELP lustre_inodes_maximum The maximum number of inodes (objects) the filesystem can hold.
# TYPE lustre_inodes_maximum gauge
lustre_inodes_maximum{component="mdt",target="fs-MDT0000",otel_scope_name="lustre"} 1885696
lustre_inodes_maximum{component="mgt",target="MGS",otel_scope_name="lustre"} 32768
lustre_inodes_maximum{component="ost",target="fs-OST0000",otel_scope_name="lustre"} 40960
lustre_inodes_maximum{component="ost",target="fs-OST0001",otel_scope_name="lustre"} 40960
# HELP lustre_available_kilobytes Number of kilobytes readily available in the pool.
# TYPE lustre_available_kilobytes gauge
lustre_available_kilobytes{component="mdt",target="fs-MDT0000",otel_scope_name="lustre"} 2365504
lustre_available_kilobytes{component="mgt",target="MGS",otel_scope_name="lustre"} 463060
lustre_available_kilobytes{component="ost",target="fs-OST0000",otel_scope_name="lustre"} 4037332
lustre_available_kilobytes{component="ost",target="fs-OST0001",otel_scope_name="lustre"} 4037332
# HELP lustre_free_kilobytes Number of kilobytes allocated to the pool.
# TYPE lustre_free_kilobytes gauge
lustre_free_kilobytes{component="mdt",target="fs-MDT0000",otel_scope_name="lustre"} 2599612
lustre_free_kilobytes{component="mgt",target="MGS",otel_scope_name="lustre"} 489272
lustre_free_kilobytes{component="ost",target="fs-OST0000",otel_scope_name="lustre"} 4106144
lustre_free_kilobytes{component="ost",target="fs-OST0001",otel_scope_name="lustre"} 4106144
# HELP lustre_capacity_kilobytes Capacity of the pool in kilobytes.
# TYPE lustre_capacity_kilobytes gauge
lustre_capacity_kilobytes{component="mdt",target="fs-MDT0000",otel_scope_name="lustre"} 2602832
lustre_capacity_kilobytes{component="mgt",target="MGS",otel_scope_name="lustre"} 491092
lustre_capacity_kilobytes{component="ost",target="fs-OST0000",otel_scope_name="lustre"} 4108388
lustre_capacity_kilobytes{component="ost",target="fs-OST0001",otel_scope_name="lustre"} 4108388
# HELP lustre_exports_total Total number of times the pool has been exported.
# TYPE lustre_exports_total gauge
lustre_exports_total{component="mdt",target="fs-MDT0000",otel_scope_name="lustre"} 32
lustre_exports_total{component="mgt",target="MGS",otel_scope_name="lustre"} 9
lustre_exports_total{component="ost",target="fs-OST0000",otel_scope_name="lustre"} 8
lustre_exports_total{component="ost",target="fs-OST0001",otel_scope_name="lustre"} 8
# HELP lustre_exports_dirty_total Total number of exports that have been marked dirty.
# TYPE lustre_exports_dirty_total gauge
lustre_exports_dirty_total{component="ost",target="fs-OST0000",otel_scope_name="lustre"} 0
lustre_exports_dirty_total{component="ost",target="fs-OST0001",otel_scope_name="lustre"} 0
# HELP lustre_exports_granted_total Total number of exports that have been marked granted.
# TYPE lustre_exports_granted_total gauge
lustre_exports_granted_total{component="ost",target="fs-OST0000",otel_scope_name="lustre"} 272832
lustre_exports_granted_total{component="ost",target="fs-OST0001",otel_scope_name="lustre"} 272832
# HELP lustre_exports_pending_total Total number of exports that have been marked pending.
# TYPE lustre_exports_pending_total gauge
lustre_exports_pending_total{component="ost",target="fs-OST0000",otel_scope_name="lustre"} 0
lustre_exports_pending_total{component="ost",target="fs-OST0001",otel_scope_name="lustre"} 0
# HELP lustre_lock_contended_total Number of contended locks.
# TYPE lustre_lock_contended_total gauge
lustre_lock_contended_total{component="mdt",target="fs-MDT0000",otel_scope_name="lustre"} 32
lustre_lock_contended_total{component="ost",target="fs-OST0000",otel_scope_name="lustre"} 32
lustre_lock_contended_total{component="ost",target="fs-OST0001",otel_scope_name="lustre"} 32
# HELP lustre_lock_contention_seconds_total Time in seconds during which locks were contended.
# TYPE lustre_lock_contention_seconds_total gauge
lustre_lock_contention_seconds_total{component="mdt",target="fs-MDT0000",otel_scope_name="lustre"} 2
lustre_lock_contention_seconds_total{component="ost",target="fs-OST0000",otel_scope_name="lustre"} 2
lustre_lock_contention_seconds_total{component="ost",target="fs-OST0001",otel_scope_name="lustre"} 2
# HELP lustre_connected_clients Number of connected clients.
# TYPE lustre_connected_clients gauge
lustre_connected_clients{component="mdt",target="fs-MDT0000",otel_scope_name="lustre"} 1
# HELP lustre_lock_count_total Number of locks.
# TYPE lustre_lock_count_total gauge
lustre_lock_count_total{component="mdt",target="fs-MDT0000",otel_scope_name="lustre"} 99
lustre_lock_count_total{component="ost",target="fs-OST0000",otel_scope_name="lustre"} 0
lustre_lock_count_total{component="ost",target="fs-OST0001",otel_scope_name="lustre"} 0
# HELP lustre_lock_timeout_total Number of lock timeouts.
# TYPE lustre_lock_timeout_total counter
lustre_lock_timeout_total{component="mdt",target="fs-MDT0000",otel_scope_name="lustre"} 0
lustre_lock_timeout_total{component="ost",target="fs-OST0000",otel_scope_name="lustre"} 0
lustre_lock_timeout_total{component="ost",target="fs-OST0001",otel_scope_name="lustre"} 0
# HELP lustre_oss_ost_stats OSS ost stats.
# TYPE lustre_oss_ost_stats gauge
lustre_oss_ost_stats{operation="obd_ping",units="usec",otel_scope_name="lustre"} 8
lustre_oss_ost_stats{operation="ost_connect",units="usec",otel_scope_name="lustre"} 60
lustre_oss_ost_stats{operation="ost_create",units="usec",otel_scope_name="lustre"} 32
lustre_oss_ost_stats{operation="ost_disconnect",units="usec",otel_scope_name="lustre"} 2
lustre_oss_ost_stats{operation="ost_get_info",units="usec",otel_scope_name="lustre"} 16
lustre_oss_ost_stats{operation="req_active",units="reqs",otel_scope_name="lustre"} 118
lustre_oss_ost_stats{operation="req_qdepth",units="reqs",otel_scope_name="lustre"} 118
lustre_oss_ost_stats{operation="req_timeout",units="sec",otel_scope_name="lustre"} 118
lustre_oss_ost_stats{operation="req_waittime",units="usec",otel_scope_name="lustre"} 118
lustre_oss_ost_stats{operation="reqbuf_avail",units="bufs",otel_scope_name="lustre"} 247
# HELP lustre_oss_ost_create_stats OSS ost_create stats.
# TYPE lustre_oss_ost_create_stats gauge
lustre_oss_ost_create_stats{operation="ost_statfs",units="usec",otel_scope_name="lustre"} 21644
lustre_oss_ost_create_stats{operation="req_active",units="reqs",otel_scope_name="lustre"} 21644
lustre_oss_ost_create_stats{operation="req_qdepth",units="reqs",otel_scope_name="lustre"} 21644
lustre_oss_ost_create_stats{operation="req_timeout",units="sec",otel_scope_name="lustre"} 21644
lustre_oss_ost_create_stats{operation="req_waittime",units="usec",otel_scope_name="lustre"} 21644
lustre_oss_ost_create_stats{operation="reqbuf_avail",units="bufs",otel_scope_name="lustre"} 43782
# HELP lustre_stats_total Number of operations the filesystem has performed.
# TYPE lustre_stats_total counter
lustre_stats_total{component="mdt",operation="getattr",target="fs-MDT0000",otel_scope_name="lustre"} 4
lustre_stats_total{component="mdt",operation="statfs",target="fs-MDT0000",otel_scope_name="lustre"} 9466
# EOF
